### PR TITLE
gen4: Derived tables

### PR DESCRIPTION
**Pull request by systay**
## Description
Adds support for derived tables in gen4.

## Related Issue(s)
#7280

## Checklist
- [x] Tests were added or are not required
- [x] Documentation was added or is not required

### CODE REVIEW

File: go/vt/vtgate/planbuilder/jointree.go [hunk 1]
```diff
@@ -102,22 +91,62 @@ type (
...
+
+		alias string
+
+		query *sqlparser.Select
+	}
+)
+
+type (
+	outerTable struct {
+		right relation
```
Comment (frouioui) [inline go/vt/vtgate/planbuilder/jointree.go:133]: Why did we decide to create 3 type declaration block?

File: go/vt/vtgate/planbuilder/jointree.go [hunk 2]
```diff
@@ -171,6 +253,46 @@ func (p parenTables) tableID() semantics.TableSet {
 	return res
 }
 
+func visitTables2(r relation, f func(tbl relation) (bool, error)) error {
```
Comment (frouioui) [inline go/vt/vtgate/planbuilder/jointree.go:256]: I this some secondary visitTables that must be done after calling `visitTables` at least once?

**Review by frouioui — APPROVED** [review]
The pull request looks good to me, I have left a couple of comments

File: go/vt/vtgate/planbuilder/jointree.go (hunk 1 above)
Comment (systay) [reply go/vt/vtgate/planbuilder/jointree.go:133]: I'll add a comment to make it clear. I collected the interface and the structs that implement it in the first two blocks

File: go/vt/vtgate/planbuilder/jointree.go (hunk 2 above)
Comment (systay) [reply go/vt/vtgate/planbuilder/jointree.go:256]: oops. forgot to clean this up

File: go/vt/vtgate/planbuilder/testdata/from_cases.txt [hunk 3]
```diff
@@ -1621,8 +1622,27 @@ Gen4 plan same as above
...
+  "Original": "select t.id from (select id from user where id = 5) as t join user_extra on t.id = user_extra.user_id",
+  "Instructions": {
+    "OperatorType": "Route",
+    "Variant": "SelectEqualUnique",
+    "Keyspace": {
+      "Name": "user",
+      "Sharded": true
+    },
+    "FieldQuery": "select t.id from (select 1 from `user` where 1 != 1), user_extra where 1 != 1",
+    "Query": "select t.id from (select 1 from `user` where id = 5), user_extra where t.id = user_extra.user_id",
```
Comment (systay) [inline go/vt/vtgate/planbuilder/testdata/from_cases.txt:1636]: select 1 from `user` where id = 5 is wrong

File: go/vt/vtgate/semantics/semantic_state.go [hunk 4]
```diff
@@ -97,47 +108,65 @@ type (
...
+
+		for _, info := range cols {
+			if col.Name.String() == info.Name {
+				return &ts, &info.Type, nil
+			}
+		}
+
+		if authoritative {
+			// if we are authoritative and we can't find the column, we should fail
+			return nil, nil, fmt.Errorf("table not found")
```
Comment (harshit-gangal) [inline go/vt/vtgate/semantics/semantic_state.go:156]: nit: we should return vterrors

File: go/vt/vtgate/planbuilder/testdata/onecase.txt [hunk 5]
```diff
@@ -1 +1,19 @@
 # Add your test case here for debugging and run go test -run=One.
+# derived table missing columns
+"select t.id from (select id from user) as t join user_extra on t.id = user_extra.user_id where t.col = 42"
+{
```
Comment (harshit-gangal) [inline go/vt/vtgate/planbuilder/testdata/onecase.txt]: missed to remove from onecase

File: go/vt/vtgate/planbuilder/jointree.go [hunk 6]
```diff
@@ -130,12 +159,65 @@ type (
...
+func (d *derivedPlan) clone() joinTree {
+	other := *d
+	return &other
+}
+
+func (d *derivedPlan) pushOutputColumns(names []*sqlparser.ColName, _ *semantics.SemTable) (offsets []int) {
+	for _, name := range names {
+		offset, err := d.findOutputColumn(name)
+		if err != nil {
+			panic(err)
```
Comment (harshit-gangal) [inline go/vt/vtgate/planbuilder/jointree.go:179]: change to return an error?

File: go/vt/vtgate/planbuilder/jointree.go [hunk 7]
```diff
@@ -130,12 +159,65 @@ type (
...
+func (d *derivedPlan) findOutputColumn(name *sqlparser.ColName) (int, error) {
+	for j, exp := range d.query.SelectExprs {
+		ae, ok := exp.(*sqlparser.AliasedExpr)
+		if !ok {
+			return 0, vterrors.New(vtrpcpb.Code_INTERNAL, "what")
+		}
+		if ae.As.IsEmpty() {
+			col, ok := ae.Expr.(*sqlparser.ColName)
+			if !ok {
+				return 0, vterrors.New(vtrpcpb.Code_INTERNAL, "what")
```
Comment (harshit-gangal) [inline go/vt/vtgate/planbuilder/jointree.go:195]: this would need a better error message.

Comment (systay) [reply go/vt/vtgate/planbuilder/jointree.go:196]: what?

**Comment by systay** [comment]
making it draft again, because: I think we need to translate the external column names before pushing down predicates, before we merge this PR

File: go/vt/vtgate/planbuilder/abstract/derived.go [hunk 8]
```diff
@@ -37,22 +37,21 @@ func (d *Derived) TableID() semantics.TableSet {
 
 // PushPredicate implements the Operator interface
 func (d *Derived) PushPredicate(expr sqlparser.Expr, semTable *semantics.SemTable) error {
-	set := semTable.ExprDeps.Dependencies(expr)
-	infoFor, err := semTable.TableInfoFor(set)
+	tableInfo, err := semTable.TableInfoFor(semTable.ExprDeps.Dependencies(expr))
```
Comment (systay) [inline go/vt/vtgate/planbuilder/abstract/derived.go:40]: this doesn't look good. we are pulling out too much data from semTable. We probably should add a method to semtable instead

Comment (frouioui) [reply go/vt/vtgate/planbuilder/abstract/derived.go]: Changed via 16d65e0

**Review by frouioui — APPROVED** [review]

File: go/vt/vtgate/planbuilder/abstract/derived.go [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+
+package abstract
+
+import (
+	"vitess.io/vitess/go/vt/sqlparser"
+	"vitess.io/vitess/go/vt/vtgate/semantics"
+)
+
+// Derived represents a derived table in the query
+type Derived struct {
```
Comment (GuptaManan100) [inline go/vt/vtgate/planbuilder/abstract/derived.go:25]: This is a new type of `Operator`. We should update the comment on `Operator` to also include this in it.

**Review by GuptaManan100 — CHANGES_REQUESTED** [review]

File: go/vt/vtgate/planbuilder/abstract/derived.go (hunk 9 above)
Comment (frouioui) [reply go/vt/vtgate/planbuilder/abstract/derived.go:25]: Updated by db656cf

File: go/vt/vtgate/planbuilder/testdata/from_cases.txt [hunk 10]
```diff
@@ -1750,6 +1830,124 @@ Gen4 plan same as above
...
+  "Original": "select t.id from (select id, textcol1 as baz from route1) as t join (select id, textcol1+textcol1 as baz from user) as s ON t.id = s.id WHERE t.baz = '3' AND s.baz = '3'",
+  "Instructions": {
+    "OperatorType": "Route",
+    "Variant": "SelectScatter",
+    "Keyspace": {
+      "Name": "user",
+      "Sharded": true
+    },
+    "FieldQuery": "select t.id from (select id, textcol1 as baz from `user` as route1 where 1 != 1) as t join (select id, textcol1 + textcol1 as baz from `user` where 1 != 1) as s on t.id = s.id where 1 != 1",
+    "Query": "select t.id from (select id, textcol1 as baz from `user` as route1) as t join (select id, textcol1 + textcol1 as baz from `user`) as s on t.id = s.id where t.baz = '3' and s.baz = '3'",
```
[the remark this answers is not in the document]
Comment (frouioui) [reply go/vt/vtgate/planbuilder/testdata/from_cases.txt:1899]: `route1` is a routed table that routes to `user`, is that a problem if we change the query to be `user as route1`?

File: go/vt/vtgate/planbuilder/testdata/from_cases.txt [hunk 11]
```diff
@@ -1621,8 +1622,27 @@ Gen4 plan same as above
...
+  "Original": "select t.id from (select id from user where id = 5) as t join user_extra on t.id = user_extra.user_id",
+  "Instructions": {
+    "OperatorType": "Route",
+    "Variant": "SelectEqualUnique",
+    "Keyspace": {
+      "Name": "user",
+      "Sharded": true
+    },
+    "FieldQuery": "select t.id from (select id from `user` where 1 != 1), user_extra where 1 != 1",
+    "Query": "select t.id from (select id from `user` where id = 5), user_extra where t.id = user_extra.user_id",
```
[the remark this answers is not in the document]
Comment (frouioui) [reply go/vt/vtgate/planbuilder/testdata/from_cases.txt:1636]: Correct, on it

File: go/vt/vtgate/planbuilder/testdata/from_cases.txt (hunk 10 above)
[the remark this answers is not in the document]
Comment (GuptaManan100) [reply go/vt/vtgate/planbuilder/testdata/from_cases.txt:1899]: no, i didn't check the routing rules, this is correct

**Review by GuptaManan100 — APPROVED** [review]
LGTM now! 🏎️

**Review by harshit-gangal — APPROVED** [review]

**Pull request merged**